Add command pipe support through :pipe option

NickLaMuro opened the pull request:
Allows for specifying the `:pipe` option in `.run`/`.run!`, which can be used to pass nested pipes to the command that is parsed in the same fashion as the base command.  This also allows for piped commands to go through the same sanitation helpers in the `CommandLineBuilder` as the base command, though passing an unfiltered string is also supported.

This feature is intended to be used to support file splitting for dumps (maybe backups?) in ManageIQ/manageiq-appliance_console, allowing us to pass `:pipe` to `PostgresAdmin` for doing things like:

```console
$ pg_dump --format c --file my_pg.dump vmdb_production | split -a 5 -b 100m - /path/to/save/my_pg.dump.
# Creates the following files for a 400MB example dump:
#   /path/to/save/my_pg.dump.aaaaa
#   /path/to/save/my_pg.dump.aaaab
#   /path/to/save/my_pg.dump.aaaac
#   /path/to/save/my_pg.dump.aaaad
```

By doing something like:

```ruby
params = {
  :format => "c",
  :file   => "my_pg.dump",
  nil     => "vmdb_production"
}

pipe   = [
  "split",
  {
    :params => [[:a, '5'], [:b, "100m"], [nil, "-"], [nil, "/path/to/save/my_pg.dump"] }
  }
]

AwesomeSpawn.run!("pg_dump", :params => params, :pipe => pipe)
```


Testing/QA
----------

I have been testing this in practice by running the steps outlined in the following gist:

https://gist.github.com/NickLaMuro/8438015363d90a2d2456be650a2b9bc6

NickLaMuro commented:
Geez coveralls... chill out...

@Fryguy going to punt this to you to start, but you might want to rope in @carbonin to get his thoughts on this as well.

What I am trying to avoid (guess I could add this to the description), is when splitting files for uploading, having to write the file to disk locally or on the remote disk, then split it (which duplicates it).  Not sure if I can pipe from `split` out to another command (which I would need to do for FTP using `curl`), but we will just have to see I guess...

carbonin commented:
I'm okay with this. But I don't have write access on this gem.

@miq-bot assign @Fryguy

Fryguy commented:
Wow this is really interesting...I need to churn this a bit in my brain.

Fryguy commented on lib/awesome_spawn.rb:
```diff
@@ -123,8 +136,14 @@ def launch(env, command, spawn_options)
+
+    if pipe = options.delete(:pipe)
+      pipe = Array(pipe)
+      cmd_str << " | " << parse_command_options(pipe[0], (pipe[1] || {}))[1]
```
I've tried to keep all command line building in the CommandLineBuilder class, so, at a minumum this should live over there.  Also, not a fan of the `[0]` `[1]` stuff...would be better to find a way to make those named variables for readability...i.e. instead of `pipe = Array(pipe)`, prefer `cmd, params = Array(pipe)`

NickLaMuro replied:
> I've tried to keep all command line building in the CommandLineBuilder class,

@Fryguy So kind of the main reason I avoided that a didn't want to pollute that class with the `:pipe` option, since there might be some command out there that takes `--pipe` as a flag, and it seemed like I would need to do some heavy duty surgery to handle that properly in that class.

That said, not opposed to doing it that way, but that is the main rational (at least from what I remember doing this a few weeks ago) for why I didn't use that class to do this.

> Also, not a fan of the `[0]` `[1]` stuff...

That is fair.  I will look in doing that, as your suggestion looks much better.  I also have a unreasonable urge to keep lines to 80 chars, so that was part of the madness that probably led to the `[0]` and `[1]` stuffs.

Fryguy replied:
If "--pipe" were a flag it would have to live inside of the :params key anyway.  Putting the :pipe key up near :env and :params would be more of a potential conflict with Open3#capture3, since we passthrough every other option at that level to capture3.  We've already taken that stance with `:params` and we've been ok, so I'm ok with also taking that stance with `:pipe`.  If Open3#capture3 adds its own pipe option, then we can deal with it then.

Fryguy commented:
In general I like the idea.  I think I'd prefer the interface to be slightly different though, in 2 ways.

1. Under the covers we use Open3.capture3 for the execution layer.  Open3 already has a method called [pipeline](https://ruby-doc.org/stdlib-2.4.3/libdoc/open3/rdoc/Open3.html#method-c-pipeline), that I think we should probably leverage instead of building via `|`.  However, this might conflict with the reasons we use capture3, so I'm not sure.

2. I picture piping as completely independent things next to each other sort of like an Array, but this interface creates a sort of visual nesting instead.

Your example, in a more visual way (not valid syntax; just for show)

```
"echo", :pipe => [
  "cat", :pipe => [
    "wc", :pipe => [
      "tr"
    ]
  ]
]
```

But I'm thinking more of a linear visualization:

```
"echo", :pipe => ["cat", "wc", "try"]
```

or with some actual code:

```
AwesomeSpawn.run('echo hello world', :pipe => [
  "cat",
  ["wc", {:params => [:c]}],
  ["tr", {:params => {:d => " "}]
])
```

NickLaMuro commented:
> 1. Under the covers we use Open3.capture3 for the execution layer. Open3 already has a method called [pipeline](https://ruby-doc.org/stdlib-2.4.3/libdoc/open3/rdoc/Open3.html#method-c-pipeline)...

Cool, I will look into that.  I kinda figured I was probably re-inventing the wheel a bit with this, and it is good to see that there is some pipeing functionality already there.  I will see what I can leverage with it.

> But I'm thinking more of a linear visualization

Oh, I like that a lot better!  And it makes much more sense since we are dealing with a ordered set of operations, and not a nested set of multiple children.


* * *

Been a bit since I have looked at this code, but I am guessing this might end up being a bit of a re-write of this.  I like the direction though, but I expect that this will take me a bit of time to make the requested changes.

Thanks for the review!

NickLaMuro commented:
@Fryguy Okay... this just got a bit more complicated.

I guess the main things of note:

* I had to write my own `pipe_capture3` since the best we could get for output and status from one of the 4 `Open3.pipeline*` methods was `.pipeline_r`, and that only returned `stdout` from the last pipe.  Documented most of were things came from, but just a heads up there.
* The `AwesomeSpawn` module now does `extend Open3` instead of calling `Open3.capture3` directly.  This allows me to use the private method `pipeline_run` in `pipe_capture3`, saving having to use send.
* I updated the interface to allow it to work like it used to, but then support piping in what seemed like intuitive ways.  This is tested pretty extensively in `#normalize_run_opts`, but these are some of the variations (basically allowing you to interchange between the splat and array syntax):
  - `AwesomeSpawn.run("cmd1", {})`
  - `AwesomeSpawn.run("cmd1", "cmd2", {})`
  - `AwesomeSpawn.run("cmd1", ["cmd2", {:params =>{}}], :env => {})`
  - `AwesomeSpawn.run(["cmd1", ["cmd2", {:params =>{}}], "cmd3"], :env => {})`

I think those are the big changes.  Let me know what you think.

(also yes, I will address those `rubocop` issues)

NickLaMuro commented:
Also, forgot to mention, I left the old version available in a branch for comparison:

https://github.com/ManageIQ/awesome_spawn/compare/master...NickLaMuro:add_pipe_option_old

Fryguy commented on lib/awesome_spawn.rb:
```diff
@@ -116,15 +138,94 @@ def build_command_line(command, params = nil)
     env = options.delete(:env) || {}
 
-    [env, build_command_line(command, params), options]
+    cmds = if commands.length == 1
```
Instead of conditionals all over with special case handling, can we wrap single commands in an array and just run everything through that?

NickLaMuro replied:
Yeah, I couldn't think of a good way to do this, and all that I came up ended up being more complex than I was hoping.

Happy to take suggestions (can be pseudo code), but at the moment I don't have a better way to implement this off the top of my head, or specifics of things that you would like me to fix.  Still agree it is a bit kludgey under the covers.

NickLaMuro replied:
The problem with this suggestion is that it causes the specs to fail, with something like this:

```console
  12) AwesomeSpawn.run! result contains #command_line
      Failure/Error: expect(subject.send(run_method, "echo", :params => %w(x)).command_line).to eq("echo x")

        expected: "echo x"
             got: ["echo x"]

        (compared using ==)
      Shared Example Group: "executes" called from ./spec/awesome_spawn_spec.rb:368
      # ./spec/awesome_spawn_spec.rb:168:in `block (4 levels) in <top (required)>
```

The change from an `String` result to an `Array` means that when this is passed into `AwesomeSpawn#launch`, everything will run through `Open3.pipe_capture3`, which I don't want.  I preferably want (and I assume others as well) to only use that patch when a piped function is being passed, otherwise use the more battle tested and less involved `Open3.capture3` instead.

Fryguy commented on lib/awesome_spawn.rb:
```diff
@@ -116,15 +138,94 @@ def build_command_line(command, params = nil)
+      pipe_in.close
+      # End block excerpt from Open3#capture3
+      [out_reader.value, err_reader.value, child_pids.last.value]
+    end
```
Cool...I wonder if you can contribute back to Ruby core.  This seems really useful.

Fryguy commented:
I don't like the `extend Open3` bit though.  It adds pubic methods to the interface that we really don't want to expose.  In this case, I think an extension to the Open3 module is more appropriate, especially if we want to push it upstream.  That extension could live here or in more_core_extensions

NickLaMuro commented:
> I don't like the extend Open3 bit though. It adds pubic methods to the interface that we really don't want to expose.

Oddly enough, that isn't the case:

```ruby
> AwesomeSpawn.methods.grep /run/
=> [:run, :run!]
irb> AwesomeSpawn.private_methods.grep /capture/
#=> [:capture3, :capture2, :capture2e, :pipe_capture3]
irb> AwesomeSpawn.capture3
NoMethodError: private method `capture3' called for AwesomeSpawn:Module
...
irb> AwesomeSpawn.pipeline_run
NoMethodError: private method `capture3' called for AwesomeSpawn:Module
...
irb> AwesomeSpawn.run
ArgumentError: no commands
...
```

Which kind of surprised me too.  Might be because they use `module_function` under the hood.  Unsure.

> In this case, I think an extension to the Open3 module is more appropriate, especially if we want to push it upstream. That extension could live here or in more_core_extensions.

The above said, I am not opposed to doing this instead either.  I think I would prefer it to live in here instead of including an gem dependency when we don't have one currently.

... but I might be biased since I am usually gem dependency adverse when ever possible.

kbrock commented on lib/core_ext/open3.rb:
```diff
@@ -0,0 +1,52 @@
+  #
+  # Should have the same interface as `Open3.capture3`, but supports passing in
+  # pipes.
+  def pipe_capture3(env, cmds, stdin_data: '', binmode: false, **opts)
```
Can we define this method in our own classe instead of monkey patching a core ruby class?

Or is there a hidden method that we need to use that I'm not seeing?

NickLaMuro replied:
> Or is there a hidden method that we need to use that I'm not seeing?

`pipeline_run` later on in this method is what we are using, and that is a private method in the `Open3` module.

kbrock commented on lib/awesome_spawn.rb:
```diff
@@ -67,9 +89,8 @@ def logger
-  def run(command, options = {})
-    bad_keys = (options.keys.flatten & [:in, :out, :err]).map { |k| ":#{k}" }
-    raise ArgumentError, "options cannot contain #{bad_keys.join(", ")}" if bad_keys.any?
+  def run(*opts)
```
would we want a different method? `run_multiple` or `run_piped`?

NickLaMuro replied:
Unless you are thinking of calling out to either `run_multiple`/`run_piped` from the method when necessary, I would say no to this suggestion.  This is a public method, and one of the goals of this PR was to not muck with the existing way the public interface worked (hence a lot of what I was testing in my specs).

NickLaMuro commented on lib/awesome_spawn.rb:
```diff
@@ -56,6 +57,27 @@ def logger
+  #   result2.output       #=> "12\n"
+  #   result2.err          #=> ""
+  #   result2.exit_status  #=> 0
+  #   result2.command_line #=> ["echo $HELLO_WORLD", "cat", "wc -c", "tr - \\\"\\ \\\""]
```
Correct, will fix.

Fryguy commented on spec/awesome_spawn_spec.rb:
```diff
@@ -15,6 +15,51 @@
+    it "supports pipes with :params" do
+      cmds = [
+        "true",
+        ["true", :params => { :arg1 => "one", :"no-arg2" => nil }]
```
Minor, but that `:"no-arg2"` can just be `:no_arg2`

NickLaMuro replied:
Ah, that is right.  I will make the change.

NickLaMuro replied:
"Done goof'd" with a typo on the first pass at fixing this, but this is all fixed now.

kbrock commented on lib/core_ext/open3.rb:
```diff
@@ -0,0 +1,52 @@
+        pipe_out.binmode
+        pipe_err.binmode
+      end
+      out_reader = Thread.new { pipe_out.read }
```
do we need to worry about joining/closing these two threads?

NickLaMuro replied:
My initial reaction is "no, because I am pretty sure this code is ripped directly from Ruby core" (I basically re-arranged it), but let me double check...

NickLaMuro replied:
Okay, so two things:

1.  This entire section is a ripoff of [`Open3#capture3`](https://github.com/ruby/ruby/blob/0a03394b44d220e5b6573e1a9542ede6e77e3bfe/lib/open3.rb#L259-L273), only thing that is different is the last section.
2.  I recently learned this, but `Thread.value` is very similar to `Thread.join`, so that is doing the `.wait` for us.

NickLaMuro commented on lib/awesome_spawn.rb:
```diff
@@ -116,15 +138,44 @@ def build_command_line(command, params = nil)
+                          [opts, {}]
+                        end
+
+    if commands.all? { |cmd| cmd.kind_of?(Array) }
```
Just noticed I need to make one addition here for the case of when all commands in the pipe are Array's with their own params hash/"tuple array".

```ruby
# Example (doesn't work right with #parse_command_options)
normalize_run_opts [["true", {:params => {:a => nil}], ["false", {:params => [:b, nil]}]]
#=> [["true", {:params => {:a => nil}, "false", {:params => [:b, nil]}], {}]
```

NickLaMuro commented:
Okay, I think I have addressed my concern with the update.  I rebased in the change into the last commit, so I can explain what I did change if the new comments and specs are not obvious.
